### PR TITLE
[FIX] mail: discuss public channel guest notification settings

### PR DESCRIPTION
**Pull request by akap-odoo**
**Current behavior before PR:**

In discuss guest user view, notification settings button was visible, which used to throw a "Session Expired" error upon being clicked.

**Desired behavior after PR is merged:**

In discuss guest user view, notification settings button is removed.

**task:3975859**

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mail/static/tests/tours/discuss_channel_as_guest_tour.js [hunk 1]
```diff
@@ -3,26 +3,38 @@
...
+        {
+            content: "Notification Settings are not available to guest users.",
+            trigger: ".o-mail-Discuss-header",
+            run() {
+                const notificationBtn = document.querySelector("[title='Notification Settings']");
+                if (notificationBtn) {
+                    throw new Error(
+                        "Notification Settings button should not be visible to guest users"
+                    );
+                }
```
Comment (rapr-odoo) [inline addons/mail/static/tests/tours/discuss_channel_as_guest_tour.js:36]: You could use `contains` instead of this.

**Review by rapr-odoo — COMMENTED** [review]
Fix looks good to me.
Linter is adding lot of unnecessary diff in the tour. So, disable the linter for that file and then update the code.

File: addons/mail/static/tests/tours/discuss_channel_as_guest_tour.js [hunk 2]
```diff
@@ -24,5 +25,15 @@ registry.category("web_tour.tours").add("discuss_channel_as_guest_tour.js", {
...
+            {
+                content: "Notification Settings are not available to guest users.",
+                trigger: ".o-mail-Discuss-header",
+                async run() {
+                    await contains(".o-mail-Discuss-header");
+                    await contains(".o-mail-Discuss-header button[title='Notification Settings']", {
+                        count: 0,
+                    });
+                },
+            },
```
Comment (alexkuhn) [inline addons/mail/static/tests/tours/discuss_channel_as_guest_tour.js:37]: I get it's good for changes to being tested, but most assertions of a negative are not reliable enough.
What I mean is that if the button is actually present but this matches another selector (e.g. `.o-mail-Discuss-topPanel a[aria-label='Change Notifications']`, then the current tour passes when we expect it shouldn't).

In general we solve this problem by asserting the positive and negative using the same selector in the same test, so that we ensure when we adapt one the other is also adapted to fully match it. This is challenging to do with how the test and feature is designed: the tour is run as a guest, and asserting the positive requires being authenticated.

We could put such positive assertion in a tour specific to an authenticated user. Problem is that this needs some "code cleaning" of the test and also we must try to keep the positive and negative assertions as close as possible in terms of locality of concern. What I mean is:

```js
// NOT OK:
// tour1.js
await contains("[title='Notification Settings']");
// tour2.js
await contains("[title='Notification Settings']", { count: 0 });

// OK:
// tour1.js
await contains(notifSettingsSelector);
// tour2.js
await contains(notifSettingsSelector, { count: 0 });
// tour_utils.js
export const notifSettingsSelector = "[title='Notification Settings']";
```

If you see what I mean. This is not great in terms of having too much indirection, but if we want to give enough trust that the negative assertion is reliable enough, then we'd need to do something like this.

-----

IMO, the test is not really necessary. It's quite obvious the feature is not intended to be there (main screen of the app, just by using it you see it), and this bug doesn't really prevent using Discuss public page. So test is nice but this feels a bit overkill IMO for the amount of LOCs this requires when this isn't really crucial.

**Review by alexkuhn — APPROVED** [review]
@robodoo r+
